helita/sim/bifrost.py

```diff
@@ -488,46 +806,724 @@ def get_quantity(self, quant, *args, **kwargs):
+
+            uni = bifrost_units()
+
+            elem = quant.split('_')
+            spic = ''.join([i for i in elem[0] if not i.isdigit()])
+            lvl = ''.join([i for i in elem[0] if i.isdigit()])
+            if self.params['do_hion'] == 1 and spic[1:-1] == 'h':
+                if quant[0] == 'n':
+                    mass = 1.0
+                else:
+                    mass = uni.m_h
+                if lvl == '1':
+
+                    return mass * (self.get_var('n1') +
+                                   self.get_var('n2') + self.get_var('n3') +
+                                   self.get_var('n4') + self.get_var('n5'))
+                else:
+                    return mass * self.get_var('n6')
+            elif self.params['do_helium'] == 1 and spic[1:-1] == 'he':
+                if quant[0] == 'n':
+                    mass = 1.0
+                else:
+                    mass = uni.m_he
+                if self.verbose:
+                    print('get_var: reading nhe%s' % lvl)
+                return mass * self.get_var('nhe%s' % lvl)
+
+            else:
+                tg = self.get_var('tg')
+                r = self.get_var('r')
+                nel = self.get_var('ne') / 1e6  # 1e6 conversion from SI to cgs
+
+                if quant[0] == 'n':
+                    dens = False
+                else:
+                    dens = True
+
+                return ionpopulation(r, nel, tg, elem=spic[1:-1], lvl=lvl,
+                                     dens=dens)
+
+        elif ((quant[:3] in MODULE_QUANT) or (
+                quant[-1] in MODULE_QUANT) or (
+                quant[-1] in SQUARE_QUANT and not(quant in CYCL_RES))):
+            # Calculate module of vector quantity
+            if (quant[:3] in MODULE_QUANT):
+                q = quant[3:]
+            else:
+                q = quant[:-1]
+            if q == 'b':
+                if not self.do_mhd:
+                    raise ValueError("No magnetic field available.")
+            result = self.get_var(q + 'xc') ** 2
+            result += self.get_var(q + 'yc') ** 2
+            if not(quant[-1] in MODULE_QUANT):
+                result += self.get_var(q + 'zc') ** 2
+
+            if (quant[:3] in MODULE_QUANT) or (quant[-1] in MODULE_QUANT):
+                return np.sqrt(result)
+            elif quant[-1] in SQUARE_QUANT:
+                return result
+
         else:
             raise ValueError(('get_quantity: do not know (yet) how to '
                               'calculate quantity %s. Note that simple_var '
                               'available variables are: %s.\nIn addition, '
                               'get_quantity can read others computed variables'
-                              ' see e.g. help(self.get_quantity) for guidance'
+                              ' see e.g. self._get_quantity? for guidance'
                               '.' % (quant, repr(self.simple_vars))))
 
+    def calc_tau(self):
```

Comment:
This is really ugly and must be rewritten. First, don't use `bifrost_units()` but import constants directly from `astropy`. Second, no for loops. Nothing in here needs a for loop. Use `scipy.integrate.quadrature.cumtrapz` for tau integration. Also, no hardcoded magical numbers and ad-hoc unit conversion.